servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TaskBasedAsyncCompletableOperator.java

```diff
@@ -153,7 +174,11 @@ final void terminal(final Object terminal) {
                     // have not seen onSubscribe and there is a sequencing issue on the Subscriber. Either way we avoid
                     // looping and deliver the terminal event.
                     try {
-                        executor.execute(this::deliverSignals);
+                        if (hasOffloaded || offload.getAsBoolean()) {
+                            executor.execute(this::deliverSignals);
```

Comment:
consider adding a comment to the effect ... `no need to set hasOffloaded because this is a terminal signal`